File: LeetCode&&Job/LinkListCopy.py
# -*- coding: utf-8 -*-
'''
https://leetcode.com/problems/copy-list-with-random-pointer/

# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
'''

import hues
hues.log('Mission',42)
hues.info('Finding',42)
hues.error(41,'is not',42)
hues.warn('We are distracted...')
hues.info('Found',24)
hues.success('Close enough')

class LinkListRandom:
    def __init__(self,x):
        self.label = x
        self.next = None
        self.random = None
 
    #插入一个节点   
    def insert(self,head,val):
        node = LinkListRandom(val)
        if head == None:
            return node
        else:
            tmp = head
            while(tmp.next != None):
                tmp = tmp.next
            tmp.next = node
        return head
    
    #打印当前链表
    def printList(self,head):
        tmp = head
        while(tmp != None):
            print 'label:',tmp.label,'  random:',(tmp.random.label if tmp.random is not None else 'none')
            tmp = tmp.next
        #print '----------------------------------------------------'


class Solution:
    # @param head: A LinkList Node
    # @return: A LinkList Node 
    def copyRandomList(self, head):
        if head == None :
            return None
        
        #首先复制原来的链表（没有random指针）
        cur = head
        headCopy = LinkListRandom(cur.label)
        curCopy = headCopy
        while(cur.next != None):
            cur = cur.next
            curCopy.next = LinkListRandom(cur.label)
            curCopy = curCopy.next
        
        #然后复制random指针
        cur = head
        curCopy = headCopy
        #计算当前cur指向的random节点距离头结点之间的距离
        while(cur != None):
            dis = self.getDis(head,cur)
            if dis != 0:
                #找出cur.random指向的那个结点
                tmp = headCopy
                for i in range(dis):
                    tmp = tmp.next
                curCopy.random = tmp
            elif cur.random == head:
                curCopy.random = head
            cur = cur.next
            curCopy = curCopy.next
            
        return headCopy
    
    def getDis(self,head,cur):
        dis = 0
        random = cur.random
        while(head != random):
            dis += 1
            head = head.next
        return dis
        
        
        




if __name__ == '__main__':
    head = None
    l = LinkListRandom('null')
    for i in range(1,10):
        head = l.insert(head,i)
    #创建了三个random指针
    head.next.next.random = head.next.next.next.next
    head.next.next.next.random = head
    head.next.next.next.next.random = head.next
    s = Solution()
    
    print 'Original:\n',l.printList(head)
    result = s.copyRandomList(head)
    print 'Changed:\n',l.printList(result)

    



            
    
    

File: LeetCode&&Job/LinkListCopy.java
import java.util.Random;

/**假设有如下一个链表：
 * struct Node
 * {
 *     int value ;
 *    struct Node *next ;
 *     struct Node *random ;
 * }   
 *其中，random指向该链表的任意一个节点或者NULL，请编程实现该链表的深拷贝。    
 * 
 *Node *deepCopy (Node *head)  
 * 核心的思想是：抛开random属性不管，先复制链表结构和值。然后计算原链表的每一个节点的random属性相对于头结点的偏移量（绝对偏移量）
 * ，也就是第几个元素，那么相应的，复制的链表中的该元素的random也指向第几个元素。另外，之所以是计算绝对偏移量而不是相对偏移量
 *（也就是某节点到该节点random属性指向的节点的相对距离），是因为，相对偏移量可以是负的，如果是负的，那就不好办了，单链表不能向反方向走
 *http://www.nowcoder.com/questionTerminal/3791a1c091e74aebb1b9d7c6a097f527
 *
 *宜信编程题目
 * 2016年10月26日 上午10:06:39
 */


public class LinkListCopy {
	class Node{
		int value;
		Node next;
		Node random;
		public Node(int value){
			this.value = value;
		}
	}
	
	/**
	 * 创建一个链表
	 * @return
	 */
	public Node createList(){
		Node head = new Node(new Random().nextInt(10));
		Node tmp = head;
		for(int i = 0; i < 10; i++){
			tmp.next = new Node(new Random().nextInt(10));
			tmp = tmp.next;
		}
		head.next.next.random = head.next.next.next.next.next;
		head.next.next.next.random = head;
		
		return head;
	}
	
	/**
	 * 得到当前节点与头节点之间的距离
	 * @param head
	 * @param cur
	 * @return
	 */
	public int getDis(Node head, Node cur){
		int dis = 0;
		Node random = cur.random;
		if(random == null) return 0;
		while(head != random){
			dis++;
			head = head.next;
		}
		return dis;
	}
	
	/**
	 * 深复制链表
	 * @param head
	 * @return
	 */
	public Node deepCopy(Node head){
		//首先复制原单链表中,从创建一个新表头节点开始
		Node cur = head;
		Node headCopy = new Node(cur.value);
		//使用curCopy来生成新的单链表
		Node curCopy = headCopy;
		while(cur.next != null){
			cur = cur.next;
			curCopy.next = new Node(cur.value);
			curCopy = curCopy.next;
		}//复制单链表next域完成
		
		//复制单链表的random域
		cur = head;
		curCopy = headCopy;
		
		while(cur != null){
			//首先求得cur指向节点的random节点与原来的表头节点之间的距离
			int dis = getDis(head,cur);
			if(dis != 0){
				Node tmp = headCopy;
				for(int i = 0; i < dis; i++){
					//找出cur.random指向的那个节点
					tmp = tmp.next;
				}
				curCopy.random = tmp;
				
			}
			else if(cur.random == head){
				curCopy.random = headCopy;
			}
			cur = cur.next;
			curCopy = curCopy.next;
		}
		
		return headCopy;
		
	}
	
	/**
	 * 输出链表
	 * @param head
	 */
	public static void printList(Node head){
		String list = "";
		while(head != null){
			list += "(" + "value:" + head.value + ",random:" + (head.random == null ? "null":head.random.hashCode()) + ")--->";
			//list+=",rando:"+(head.random==null?"":head.random.hashCode())+")-->";
			head = head.next;
		}
		System.out.println(list);
	}
	
	/**
	 * 
	 * @param args
	 */
	public static void main(String[] args){
		//新建一个单链表
		LinkListCopy2 copy = new LinkListCopy2();
		Node head = copy.createList();
		Node headCopy = copy.deepCopy(head);
		printList(headCopy);
	}
	
	
}
